File: services/semantic/src/domain/models/submission.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime

SUBMISSION_STATUS_DRAFT = "draft"
SUBMISSION_STATUS_PENDING = "pending"
SUBMISSION_STATUS_APPROVED = "approved"
SUBMISSION_STATUS_REJECTED = "rejected"


@dataclass(slots=True)
class SubmissionModel:
    submission_id: int
    created_by_user_id: int
    source_identifier: str | None = None
    title: str = ""
    abstract: str = ""
    year: int | None = None
    best_oa_location: str | None = None
    status: str = "draft"
    referenced_works: list[str] = field(default_factory=list)
    related_works: list[str] = field(default_factory=list)
    moderated_by_user_id: int | None = None
    moderation_comment: str | None = None
    approved_paper_id: int | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    submitted_at: datetime | None = None
    moderated_at: datetime | None = None

__all__ = [
    "SUBMISSION_STATUS_APPROVED",
    "SUBMISSION_STATUS_DRAFT",
    "SUBMISSION_STATUS_PENDING",
    "SUBMISSION_STATUS_REJECTED",
    "SubmissionModel",
]

File: services/gateway/internal/transport/http/presenters/submission.go
package presenters

type SubmissionUpsertRequest struct {
	SourceIdentifier string   `json:"source_identifier"`
	Title            string   `json:"title"`
	Abstract         string   `json:"abstract"`
	Year             int      `json:"year"`
	BestOALocation   string   `json:"best_oa_location"`
	ReferencedWorks  []string `json:"referenced_works"`
	RelatedWorks     []string `json:"related_works"`
}

type SubmissionResponse struct {
	Submission SubmissionRecord `json:"submission"`
}

type SubmissionRecord struct {
	SubmissionID      int64    `json:"submission_id"`
	CreatedByUserID   int64    `json:"created_by_user_id"`
	SourceIdentifier  string   `json:"source_identifier"`
	Title             string   `json:"title"`
	Abstract          string   `json:"abstract"`
	Year              int      `json:"year"`
	BestOALocation    string   `json:"best_oa_location"`
	ReferencedWorks   []string `json:"referenced_works"`
	RelatedWorks      []string `json:"related_works"`
	Status            string   `json:"status"`
	ModeratedByUserID int64    `json:"moderated_by_user_id"`
	ModerationComment string   `json:"moderation_comment"`
	ApprovedPaperID   int64    `json:"approved_paper_id"`
	CreatedAt         string   `json:"created_at"`
	UpdatedAt         string   `json:"updated_at"`
	SubmittedAt       string   `json:"submitted_at"`
	ModeratedAt       string   `json:"moderated_at"`
}

type SubmissionListResponse struct {
	Items  []SubmissionRecord `json:"items"`
	Total  int64              `json:"total"`
	Limit  int64              `json:"limit"`
	Offset int64              `json:"offset"`
}

type ModerateSubmissionRequest struct {
	Action  string `json:"action"`
	Comment string `json:"comment"`
}
